core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala

```diff
@@ -0,0 +1,123 @@
+        new FakeTask(0),
+        HashMap[String, Long](),
+        HashMap[String, Long](),
+        serializer.newInstance())
+
+    // the program should run in this order:
+    // +-----------------------------+----------------------------------------------+
+    // |      main test thread       |      worker thread                           |
+    // +-----------------------------+----------------------------------------------+
+    // |    executor.launchTask()    |                                              |
+    // |                             | TaskRunner.run() begins                      |
+    // |                             |          ...                                 |
+    // |                             | execBackend.statusUpdate  // 1st time, #L240 |
+    // | executor.killAllTasks(true) |                                              |
+    // |                             |          ...                                 |
+    // |                             |  task = ser.deserialize   // #L253           |
+    // |                             |          ...                                 |
+    // |                             | execBackend.statusUpdate  // 2nd time, #L365 |
+    // |                             |          ...                                 |
+    // |                             |   TaskRunner.run() ends                      |
+    // |       check results         |                                              |
+    // +-----------------------------+----------------------------------------------+
+
+    val mockExecutorBackend = mock(classOf[ExecutorBackend])
+    when(mockExecutorBackend.statusUpdate(any(), any(), any()))
+      .thenAnswer(new Answer[Unit] {
+        var firstTime = true
+        override def answer(invocationOnMock: InvocationOnMock): Unit = {
+          if (firstTime) {
+            TestHelper.latch1.countDown()
+            // here between latch1 and latch2, executor.killAllTasks() is called
+            TestHelper.latch2.await()
+            firstTime = false
+          }
+          else {
+            val taskState = invocationOnMock.getArguments()(1).asInstanceOf[TaskState]
+            // save the returned `taskState` and `testFailedReason` into TestHelper
+            TestHelper.taskState = taskState
+            val taskEndReason = invocationOnMock.getArguments()(2).asInstanceOf[ByteBuffer]
+            TestHelper.testFailedReason = serializer.newInstance().deserialize(taskEndReason)
+            // let the main test thread to check `taskState` and `testFailedReason`
+            TestHelper.latch3.countDown()
+          }
+        }
+      })
+
+    val executor = new Executor("id", "localhost", mockEnv, userClassPath = Nil, isLocal = true)
+    // the task will be launched in a dedicated worker thread
+    executor.launchTask(mockExecutorBackend, 0, 0, "", serializedTask)
+
+    TestHelper.latch1.await()
+    executor.killAllTasks(true)
```

Comment:
add a comment above this line, something like:
```scala
// we know the task will be started, but not yet deserialized, because of the latches we use in mockBackend.
```